File: 704-binary-search/704-binary-search.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        def binsearch(left, right):
            if left > right:
                return -1
            
            mid = (left + right) // 2
            
            if nums[mid] == target:
                return mid
            elif target > nums[mid]:
                return binsearch(mid+1, right)
            
            return binsearch(left, mid-1)
        
        return binsearch(0, len(nums)-1)
        
        
        
        
        
#         return self.binsearch(nums, target, 0, len(nums)-1)
        
#     def binsearch(self, nums, target, start, end):
#         if start > end:
#             return -1
        
#         mid = (start + end) // 2
        
#         if nums[mid] == target:
#             return mid
#         elif target > nums[mid]:
#             return self.binsearch(nums, target, mid+1, end)
#         else:
#             return self.binsearch(nums, target, start, mid-1)
    
    
        

File: 704-binary-search/704-binary-search.java
class Solution {
    private static int[] nums;
    
    public int search(int[] nums, int target) {
        this.nums = nums;
        return binSearch(0, nums.length-1, target);
    }
    
    private static int binSearch(int low, int high, int target) {
        if (low > high) return -1;
        
        int mid = low + (high - low) / 2;
        int midVal = nums[mid];
        if (midVal < target) return binSearch(mid+1, high, target);
        else if (midVal > target) return binSearch(low, mid-1, target);
        
        return mid;            
    }
}